File: Algorithms/C++/Tree/107_Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> values;

        if (root == NULL) {
            return values;
        }

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            vector<int> level_values;
            int level_size = q.size();

            for (int i = 0; i < level_size; i++) {
                TreeNode* node = q.front();
                q.pop();

                level_values.push_back(node->val);

                if (node->left != NULL) {
                    q.push(node->left);
                }
                if (node->right != NULL) {
                    q.push(node->right);
                }
            }

            values.insert(values.begin(), level_values);
        }

        return values;
    }
};

File: Algorithms/Swift/Easy/107_Binary_Tree_Level_Order_Traversal_II.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func levelOrderBottom(root: TreeNode?) -> [[Int]] {
        if root == nil {
            return [[Int]]()
        }

        var reversedLevels = [[Int]]()
        var queue = [TreeNode?]()

        queue.append(root)
        while !queue.isEmpty {
            var currentLevel = [Int]()

            let currentLevelCount = queue.count
            for _ in 0..<currentLevelCount {
                let node = queue.first!
                queue.removeFirst()

                currentLevel.append(node!.val)

                if node!.left != nil {
                    queue.append(node!.left)
                }
                if node!.right != nil {
                    queue.append(node!.right)
                }
            }

            reversedLevels.append(currentLevel)
        }

        var levels = [[Int]]()

        while !reversedLevels.isEmpty {
            levels.append(reversedLevels.last!)
            reversedLevels.removeLast()
        }

        return levels
    }
}
